File: lib/src/main/scala/net/kemitix/thorp/lib/Filters.scala
package net.kemitix.thorp.lib

import java.io.File
import java.nio.file.Path

import net.kemitix.thorp.config.Configuration
import net.kemitix.thorp.domain.Filter
import net.kemitix.thorp.domain.Filter.{Exclude, Include}

import scala.jdk.CollectionConverters._

object Filters {

  def isIncluded(configuration: Configuration, file: File): Boolean =
    isIncluded(file.toPath)(configuration.filters.asScala.toList)

  def isIncluded(p: Path)(filters: List[Filter]): Boolean = {
    sealed trait State
    final case class Unknown() extends State
    final case class Accepted() extends State
    final case class Discarded() extends State
    val excluded = isExcludedByFilter(p)(_)
    val included = isIncludedByFilter(p)(_)
    filters.foldRight(Unknown(): State)(
      (filter, state) =>
        (filter, state) match {
          case (_, Accepted())                => Accepted()
          case (_, Discarded())               => Discarded()
          case (x: Exclude, _) if excluded(x) => Discarded()
          case (i: Include, _) if included(i) => Accepted()
          case _                              => Unknown()
      }
    ) match {
      case Accepted()  => true
      case Discarded() => false
      case Unknown() =>
        filters.forall {
          case _: Include => false
          case _          => true
        }
    }
  }

  def isIncludedByFilter(path: Path)(filter: Filter): Boolean =
    filter.predicate.test(path.toFile.getPath)

  def isExcludedByFilter(path: Path)(filter: Filter): Boolean =
    filter.predicate.test(path.toFile.getPath)

}

File: lib/src/main/java/net/kemitix/thorp/lib/Filters.java
package net.kemitix.thorp.lib;

import net.kemitix.thorp.config.Configuration;
import net.kemitix.thorp.domain.Filter;

import java.io.File;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

public interface Filters {

    static boolean isIncluded(
            Configuration configuration,
            File file
    ) {
        return isIncluded(file.toPath(), configuration.filters);
    }

    static boolean isIncluded(
            Path path,
            List<Filter> filters
    ) {
        AtomicBoolean included = new AtomicBoolean(isIncludedByDefault(filters));
        filters.forEach(
                filter -> {
                    if (included.get()) {
                        if (filter instanceof Filter.Exclude) {
                            boolean excludedByFilter = isExcludedByFilter(path, filter);
                            included.set(!excludedByFilter);
                        }
                    } else {
                        if (filter instanceof Filter.Include) {
                            boolean includedByFilter = isIncludedByFilter(path, filter);
                            included.set(includedByFilter);
                        }
                    }
                }
        );
        return included.get();
    }

    static boolean isIncludedByDefault(List<Filter> filters) {
        return filters.isEmpty() ||
                filters.stream()
                        .allMatch(filter ->
                                filter instanceof Filter.Exclude);
    }

    static boolean isIncludedByFilter(Path path, Filter filter) {
        return  filter.predicate().test(path.toFile().getPath());
    }
    static boolean isExcludedByFilter(Path path, Filter filter) {
        return  filter.predicate().test(path.toFile().getPath());
    }
}
